src/addr.rs

```diff
@@ -529,6 +578,22 @@ impl Sub<PhysAddr> for PhysAddr {
     }
 }
 
+/// Checks whether the address is canonical.
+
+/// Returns an `Option<u64>` containing the queried address representing if
+/// the address is canonical or not.
+#[inline]
+const fn is_canonical(addr: u64) -> Option<u64> {
+    let msb = (addr >> 47) & 0x1ffff;
+
+    // The most-significant 17 bits must all be one or all be 0.
+    if msb == 0x0 || msb == 0x1ffff {
+        Some(addr)
+    } else {
+        None
+    }
```

Comment:
We can shorten this implementation to just be:
```rust
let trunc = VirtAddr::new_truncate(addr);
if trunc.as_u64() == addr {
    Some(trunc)
} else {
    None
}
```
This implementation still generates good code ([Godbolt link](https://godbolt.org/z/o3hn83WGv));

Reply:
:+1: also kuddos on being so thorough that you provide a godbolt link on a pr review.